text/0097-log-data-model.md

```diff
@@ -0,0 +1,972 @@
+  may have certain control over what information is included, e.g. customize the
+  format. An example is Apache log file.
+
+- First-party Applications. These are applications that we develop and we have
+  some control over how the logs and events are generated and what information
+  we include in the logs. We can likely modify the source code of the
+  application if needed.
+
+### Field Kinds
+
+This Data Model defines a logical model for a log record (irrespective of the
+physical format and encoding of the record). Each record contains 2 kinds of
+fields:
+
+- Named top-level fields of specific type and meaning.
+
+- Fields stored in the key/value pair lists, which can contain arbitrary values
+  of different types. The keys and values for well-known fields follow semantic
+  conventions for key names and possible values that allow all parties that work
+  with the field to have the same interpretation of the data. See references to
+  semantic conventions for `Resource` and `Attributes` fields and examples in
+  [Appendix A](#appendix-a-example-mappings).
+
+The reasons for having these 2 kinds of fields are:
+
+- Ability to efficiently represent named top-level fields, which are almost
+  always present (e.g. when using encodings like Protocol Buffers where fields
+  are enumerated but not named on the wire).
+  
+- Ability to enforce types of named fields, which is very useful for compiled
+  languages with type checks.
+
+- Flexibility to represent less frequent data via key/value pair lists. This
+  includes well-known data that has standardized semantics as well as arbitrary
+  custom data that the application may want to include in the logs.
+
+When designing this data model I followed the following reasoning to make a
+decision about when to use use a top-level named field:
+
+- The field needs to be either mandatory for all records or be frequently
+  present in well-known log and event formats (such as `Timestamp`) or is
+  expected to be often present in log records in upcoming logging systems (such
+  as `TraceId`).
+
+- The field’s semantics must be the same for all known log and event formats and
+  can be mapped directly and unambiguously to this data model.
+  
+Both of the above conditions were required to give the field a place in the
+top-level structure of the record.
+
+## Log and Event Record Definition
+
+Note: below we use type `any`, which can be a scalar value (number, string or
+boolean), or an array or map of values. Arbitrary deep nesting of values for
+arrays and maps is allowed (essentially allow to represent an equivalent of a
+JSON object).
+
+[Appendix A](#appendix-a-example-mappings) contains many examples that show how
+existing log formats map to the fields defined below. If there are questions
+about the meaning of the field reviewing the examples may be helpful.
+
+Here is the list of fields in a log record:
+
+Field Name     |Description
```

Comment:
Consider to include https://github.com/open-telemetry/opentelemetry-proto/blob/master/opentelemetry/proto/common/v1/common.proto#L59 as part of the data model to be consistent with trace/metrics.

Reply:
It is probably going to be very rare that we need this in log record. Logs are typically coming from 2 sources:
- Legacy log sources - Syslog, Winlogs, File logs, etc. These don't have any information about instrumentation libraries.
- Newer log sources that are application logs emitted using OpenTelemetry-compliant logging libraries. Even for this I don't think we can expect any notion of "instrumentation" library that is helpful like it is for traces or metrics.

I think we should have a semantic convention for storing instrumentation library info in the `Attributes` in the rare cases that it may be needed (I really can't see people writing instrumentation libraries that add logs to other libraries. Why would it be desirable? If you go into the trouble it is better to implement a tracing instrumentation instead).

Reply:
Not a logging expert, but the first example that I have in mind is the Java `Logger`. They have an API `Logger.getLogger` where I would expect the `name` passed to that call to be the name used in the `InstrumentationLibrary`.

See the initial otep about that https://github.com/open-telemetry/oteps/blob/f4de6261d32955539ec9af0298a0203d09c13735/text/0016-named-tracers.md#explanation

Reply:
I think Java `Logger` serves a bit different purpose than the instrumentation library name as [defined in OpenTelemetry](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#obtaining-a-tracer):
>This name must identify the instrumentation library (also referred to as integration, e.g. io.opentelemetry.contrib.mongodb) and **not the instrumented library**.

I think this definition excludes certain use cases for Logger. I believe Logger name can be the name of the instrumented library (which is directly prohibited by OpenTelemetry's definition). Logger name can be also the application module name or the name of the Java class, neither of which fits OpenTelemetry's definition of "instrumentation library name".

I would suggest to leave this out from the data model until there is clarity in this topic.

Reply:
> I think this definition excludes certain use cases for Logger. I believe Logger name can be the name of the instrumented library (which is directly prohibited by OpenTelemetry's definition) 
If the instrumentation happens directly in the main library then the instrumented library is equivalent with the instrumentation library, which is mostly the case for logs.

My 2 cents on this: I think by having it in the logs will help us exercising that part of the protocol and force us to generalize it better.

Reply:
Sounds good regarding classifying java `Logger` as the instrumentation library as OpenTelemetry defines it.

However, I am still not convinced `InstrumentationLibrary` must be a top-level field for logs for a couple reasons:
1.  InstrumentationLibrary is going to be absent in legacy logs so we will waste space and more importantly see (2) below.
2. The way InstrumentationLibrary is currently defined in OTLP makes it very high cost even when it is not needed because it requires an additional nesting level for ProtoBufs. I am very reluctant to do the same thing for logs and have significant performance hit due to a field that very large portion of log messages will not use.

I may be more easily convinced if we refactor OTLP to avoid requiring a whole nesting level for InstrumentationLibrary and make it an attribute instead (e.g. "instrumentation.lbrary.name"/"instrumentation.lbrary.version" pair).

Reply:
Common practice is to use the fully qualified class name as the logger for a Java Logger, and then using the package as a hierarchy. 

So if I had a webapp, the logger might be associated with a classs `com.example.webapp.persistence.serializer.PersonSerializer`. That logger is then considered a "child logger" of `com.example.webapp.persistence`, so if a logging level is set on the parent, the child inherits that configuration. 

Repurposing the logger name has two problems- it breaks that expected behavior, and it's also making a change to the user's code- for example if we pull logs from Spring using a Logback exporter, we will get logger names based on the class name, not the instrumentation library as we expect it- only newly created code that broke with common practices would include the logging library name/version as we expect it.

It would make better sense in my opinion to make this an explicit field accessible by our internal log transmission library. That way if we instrument an application through a java agent for example, we can at that point specify the library name and version in a way that would be consistent with tracing, but logs that were pulled in from Logger or Logback or Log4J would either have an empty default, or the library name would be Log4J, etc.

Reply:
@zenmoto I think Bogdan's proposal is the following:
- Logger name stays how it is defined in Java logging libraries. Nothing changes here. It can be the class name or whatever is used in the application.
- Now imagine a separate piece of code is added to the application that uses Java logging. That piece of code is usually called "instrumenting library" in OpenTelemetry and one of the purposes of the instrumenting library (among other things) is to send logs produced by Java logging library via a network protocol (e.g. OTLP). For example the instrumenting library can add an Appender (or whatever means for extension the particular logging library supports).
- When constructing an OTLP request the Logger's "name" field is placed in the InstrumentationLibrary.name field, not the name of the "instrumenting library" that added the Appender.

That's my understanding, but I may be wrong.

Reply:
I think that for instrumentation we probably don't want to use the Java Logger, as the java.util.Logger api is very string-based- it would be difficult to build this logging record using that API.

Reply:
I posted proposals to remove InstrumentationLibrary from OTLP protocol (and hence from data model) of traces and metrics as top-level concept and make them semantic conventions instead.

Please have a look:
https://github.com/open-telemetry/opentelemetry-proto/issues/148
https://github.com/open-telemetry/opentelemetry-proto/issues/149

If this gets accepted we will apply the same logic to log data model.

Reply:
Resolving. Feel free to reopen if needed.